pkg/auth/mtls.go

```diff
@@ -0,0 +1,49 @@
+package auth
+
+import (
+	"context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+func NewMTLSAuthenticator() Authenticator {
+	return &mtlsAuthenticator{}
+}
+
+type mtlsAuthenticator struct{}
+
+// Authenticate implements Authenticator.
+func (*mtlsAuthenticator) Authenticate(ctx context.Context) (AuthenticatedUser, error) {
+	info, ok := peer.FromContext(ctx)
+	if !ok {
+		return "", status.Errorf(codes.Internal, "no peer info found")
+	}
+	if info.AuthInfo == nil {
+		return "", status.Errorf(codes.Unauthenticated, "no auth info found")
+	}
+	switch ai := info.AuthInfo.(type) {
+	case credentials.TLSInfo:
+		var subject string
+		for _, vc := range ai.State.VerifiedChains {
+			if len(vc) == 0 {
+				continue
+			}
+			leaf := vc[0]
+			cn := leaf.Subject.CommonName
+			if cn == "" {
+				continue
+			}
```

Comment:
it it safe to fallback to Config.ClientCAs   in case of missing PeerCertificates[0].CommonName ? 

> 	// VerifiedChains is a list of one or more chains where the first element is
	// PeerCertificates[0] and the last element is from Config.RootCAs (on the
	// client side) or Config.ClientCAs (on the server side).
    VerifiedChains [][]*x509.Certificate

Reply:
That continue is for the loop over the list of chains, not the certs within each chain. I am only checking the CN of the leaf cert in each chain (there may be multiple)